cpp/include/raft/neighbors/ivf_flat-inl.cuh

```diff
@@ -464,6 +464,52 @@ void search(raft::resources const& handle,
                 nullptr);
 }
 
+/**
+ * @brief Reconstruct vector database
+ *
+ *
+ * @tparam T data element type
+ * @tparam IdxT type of the indices
+ *
+ * @param[in] handle
+ * @param[in] index ivf-flat constructed index
+ * @param[in] vector_ids vector with the ids of vectors to look for
+ * @param[out] vector_out matrix with the database of vectors
+ */
+template <typename T, typename IdxT>
+void reconstruct_batch(raft::resources const& handle,
+                       const index<T, IdxT>& index,
+                       raft::device_vector_view<const IdxT, IdxT> vector_ids,
+                       raft::device_matrix_view<T, IdxT, row_major> vector_out)
+{
+  return raft::neighbors::ivf_flat::detail::reconstruct_batch(
+    handle, index, vector_ids, vector_out);
+}
+
+/**
+ * @brief Reconstruct vectors of a given cluster
+ *
+ *
+ * @tparam T data element type
+ * @tparam IdxT type of the indices
+ *
+ * @param[in] handle
+ * @param[in] index ivf-flat constructed index
+ * @param[out] vector_out matrix with the vectors contained in the cluster
+ * @param[in] label cluster index
+ * @param[in] offset offset
+ */
+template <typename T, typename IdxT>
+void reconstruct_list_data(raft::resources const& handle,
+                           const index<T, IdxT>& index,
+                           device_matrix_view<T, IdxT, row_major> out_vectors,
```

Comment:
In `ivf_pq`, list and query dimensions are limited to `uint32_t` rather than `IdxT` (type of DB indices).

Here in the ivf_flat API you use  `device_matrix_view<T, IdxT, row_major>` rather than  `device_matrix_view<T, uint32_t, row_major>` , but you also set both label and offset arguments to `uint32_t`. Could you please check here, in the overload above, and down the detail namespace what are the appropriate indexing types and make it consistent?

Reply:
We are using `IdxT` for the IVF-Flat API. I corrected the code so that it uses the template type.

Reply:
Oh, interesting! Out of curiosity, could you add a test case for ivf-flat with 64-bit indexing type and try to call something with an offset slightly outside the 32-bit range? :)

Reply:
A quick browsing confirms that 64-bit indexing/extents are actually not supported https://github.com/rapidsai/raft/blob/cad88b38a1de065fed933e8a910002d543b8cfa2/cpp/include/raft/neighbors/ivf_flat-inl.cuh#L455-L464

Reply:
The portion of code you are mentioning is indeed probably a mistake. At the moment the IVF-Flat API functions are templated with `IdxT` everywhere, and it is set to use `uint64_t` during compilation. Do you recommend that we update the IVF-Flat API to systematically use `uint32_t` instead? If so, this is an extensive change, and that would maybe require a separate PR, right?

Reply:
If ivf-flat can actually work with extents and input sizes `IdxT` larger than 32-bit, I don't see any problem using `IdxT` in the api. But I suspect this is not the case. Hence we should fix it either by limiting extent types to 32-bit or making sure it works with all possible `IdxT` (not only those we instantiate explicitly btw). I agree this work is worth a separate PR.
However, for the new public functions added in this PR, I think it makes sense to fix the api from the beginning, keeping in mind the future changes in the other parts of the code.